pkg/blockbuilder/scheduler/scheduler.go

```diff
@@ -2,43 +2,134 @@ package scheduler
-	panic("unimplemented")
+	TargetRecordConsumptionPeriod time.Duration
 }
 
-func (s unimplementedScheduler) HandleCompleteJob(_ context.Context, _ string, _ *types.Job) error {
-	panic("unimplemented")
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.DurationVar(&cfg.Interval, prefix+"interval", time.Minute, "How often to run scheduler top compute jobs.")
+	f.DurationVar(&cfg.TargetRecordConsumptionPeriod, prefix+"target-records-spanning-period", time.Hour, "Period used by the planner to calculate the start and end offset such that each job consumes records spanning the target period.")
+	f.StringVar(&cfg.ConsumerGroup, prefix+"consumer-group", "block-scheduler", "Consumer group used by block scheduler to track the last consumed offset.")
 }
 
-func (s unimplementedScheduler) HandleSyncJob(_ context.Context, _ string, _ *types.Job) error {
-	panic("unimplemented")
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("block-scheduler.", f)
+}
+
+func (cfg *Config) Validate() error {
+	if cfg.Interval.Seconds() <= 0 {
+		return errors.New("interval must be a non-zero value")
+	}
+
+	return nil
 }
 
-// QueueScheduler implements the Scheduler interface
-type QueueScheduler struct {
-	queue *JobQueue
+// BlockScheduler implements the Scheduler interface
+type BlockScheduler struct {
+	services.Service
+
+	cfg     Config
+	logger  log.Logger
+	queue   *JobQueue
+	metrics *Metrics
+
+	offsetReader OffsetReader
+	planner      Planner
 }
 
 // NewScheduler creates a new scheduler instance
-func NewScheduler(queue *JobQueue) *QueueScheduler {
-	return &QueueScheduler{
-		queue: queue,
+func NewScheduler(cfg Config, queue *JobQueue, offsetReader OffsetReader, logger log.Logger, r prometheus.Registerer) *BlockScheduler {
+	planner := NewTimeRangePlanner(cfg.TargetRecordConsumptionPeriod, offsetReader, func() time.Time { return time.Now().UTC() }, logger)
+	s := &BlockScheduler{
+		cfg:          cfg,
+		planner:      planner,
+		offsetReader: offsetReader,
+		logger:       logger,
+		metrics:      NewMetrics(r),
+		queue:        queue,
+	}
+	s.Service = services.NewBasicService(nil, s.running, nil)
+	return s
+}
+
+func (s *BlockScheduler) running(ctx context.Context) error {
+	ticker := time.NewTicker(s.cfg.Interval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := s.runOnce(ctx); err != nil {
+				// TODO: add metrics
+				level.Error(s.logger).Log("msg", "failed to schedule jobs", "err", err)
+			}
+		case <-ctx.Done():
+			return nil
+		}
 	}
 }
 
-func (s *QueueScheduler) HandleGetJob(ctx context.Context, builderID string) (*types.Job, bool, error) {
+func (s *BlockScheduler) runOnce(ctx context.Context) error {
+	lag, err := s.offsetReader.GroupLag(ctx)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "failed to get group lag", "err", err)
+		return err
+	}
+
+	if err := s.publishLagMetrics(lag); err != nil {
```

Comment:
tiny nit: this looks coupled to the `GroupLag` call -- let's instead implement a `scheduler.grouplag()` method which defers to the offset reader and afterwards updates metrics -- or just put the metrics in the `OffsetReader` itself. It's cleaner that way and more likely to be resilient to refactoring changes.